gnomad_constraint/utils/constraint.py

```diff
@@ -195,12 +217,173 @@ def create_constraint_training_dataset(
     possible_ht = possible_ht.transmute(possible_variants=possible_ht.variant_count)
 
     # Outer join the Tables with possible variant counts and observed variant counts.
-    ht = ht.join(possible_ht, "outer")
-    # Annotate the Table with cpg site and mutation_type (one of "CpG", "non-CpG
+    ht = observed_ht.join(possible_ht, "outer")
+
+    tmp_path = new_temp_file(prefix="constraint", extension="ht")
+    ht.write(tmp_path)
+    ht = hl.read_table(tmp_path)
+
+    # Annotate the Table with cpg sites and mutation_type (one of "CpG", "non-CpG
     # transition", or "transversion").
     ht = annotate_mutation_type(ht)
-    # Annotate parameters used in the function as global annotations.
+
+    if global_annotation:
+        ht = ht.annotate_globals(
+            **{global_annotation: hl.struct(max_af=max_af, pops=pops)}
+        )
+
+    return ht
+
+
+def apply_models(
+    exome_ht: hl.Table,
+    context_ht: hl.Table,
+    mutation_ht: hl.Table,
+    plateau_models: hl.StructExpression,
+    coverage_model: Tuple[float, float],
+    max_af: float = 0.001,
+    keep_annotations: Tuple[str] = (
+        "context",
+        "ref",
+        "alt",
+        "methylation_level",
+    ),
+    pops: Tuple[str] = (),
+    partition_hint_for_counting_variants: int = 2000,
+    partition_hint_for_aggregation: int = 1000,
+    custom_vep_annotation: str = None,
+    cov_cutoff: int = COVERAGE_CUTOFF,
+) -> hl.Table:
+    """
+    Compute the expected number of variants and observed:expected ratio using plateau models and coverage model.
+
+    This function sums the number of possible variants times the mutation rate for all
+    variants, and applies the calibration model separately for CpG transitions and
+    other sites. For sites with coverage lower than the coverage cutoff, the value
+    obtained from the previous step is multiplied by the coverage correction factor.
+    These values are summed across the set of variants of interest to obtain the
+    expected number of variants.
+
+    A brief view of how to get the expected number of variants:
+        mu_agg = the number of possible variants * the mutation rate (all variants)
+        predicted_proportion_observed = sum(plateau model slop * mu_agg + plateau model intercept) (separately for CpG transitions and other sites)
+        if 0 < coverage < coverage cutoff:
+            coverage_correction = coverage_model slope * log10(coverage) + coverage_model intercept
+            expected_variants = sum(predicted_proportion_observed * coverage_correction)
+        else:
+            expected_variants = sum(predicted_proportion_observed)
+        The expected_variants are summed across the set of variants of interest to
+        obtain the final expected number of variants.
+
+    Function adds the following annotations:
```

Comment:
Since all of these seem to be grouped by groupings you could just add that once here "... all grouped by groupings (output of `annotate_exploded_vep_for_constraint_groupings`)" and remove that specification below 


In other cases where you put a function name you also add `()` so maybe add that here too